posydon/binary_evol/MESA/step_mesa.py

```diff
@@ -1514,27 +1523,36 @@ def __call__(self, binary):
         # period inside the grid, but m1 outside the grid
         elif ((not self.flip_stars_before_step and
                self.p_min <= p <= self.p_max and
-                (m1 < self.m1_min or m1 > self.m1_max)
-                )):
+                (m1 < self.m1_min or m1 > self.m1_max))):
             set_binary_to_failed(self.binary)
             raise GridError(f'The mass of m1 ({m1}) is outside the grid,'
-                                'while the period is inside the grid.')
+                                ' while the period is inside the grid.')
         
         # period inside the grid, but m2 outside the grid
         elif ((not self.flip_stars_before_step and
                self.p_min <= p <= self.p_max and
-               (m2 < self.m2_min or m2 > self.m2_max)
-               )):
+               (m2 < self.m2_min or m2 > self.m2_max))):
             set_binary_to_failed(self.binary)
             raise GridError(f'The mass of m2 ({m2}) is outside the grid,'
-                                'while the period is inside the grid.')
+                                ' while the period is inside the grid.')
+        
+        # period inside the grid, but m1 outside the grid with flipped stars
+        elif ((self.flip_stars_before_step and 
+               self.p_min <= p <= self.p_max and
+               m1 < self.m2_min or m1 > self.m2_max)):
+            set_binary_to_failed(self.binary)
+            raise GridError(f'The mass of m2 ({m1}) is outside the grid,'
+                                ' while the period is inside the grid.')
 
-        else:
-            if len(self.binary.state_history) > 2:
-                if self.binary.state_history[-2] == 'detached':
-                    set_binary_to_failed(self.binary)
-                    raise GridError('CO_HeMS_RLO binary outside grid and coming from detached')
+        # period inside the grid, but m2 outside the grid with flipped stars
+        elif ((self.flip_stars_before_step and
+                self.p_min <= p <= self.p_max and
+                m2 < self.m1_min or m2 > self.m1_max)):
+            set_binary_to_failed(self.binary)
+            raise GridError(f'The mass of m1 ({m2}) is outside the grid,'
+                                ' while the period is inside the grid.')
 
+        else:
```

Comment:
In this final `else` - is it known what binaries are expected to enter here? It is hard for me to tell if all the other `elif`'s are catching the expected behavior and this is a final fail safe or if this `else` is actually catching a known case that we understand. I think just some comment explaining the intended behavior may be nice!

Reply:
I tried to cover all case where something might not work because of the associated grid. The `else`, on the other hand, covers binaries, where we can redirect them to the detached step.
There are known cases of binaries going through the redirect. The example below is from an slack post on the developers channel from Zepei. This is/was intended behaviour, although naming could be better.

![Screenshot 2024-04-11 at 08 19 32](https://github.com/POSYDON-code/POSYDON/assets/14039563/4c91f741-9106-475a-9246-39188b57ea46)

We could change the logic to manually include these redirects and catch any unexpected system with the else instead (which might be better).